src/jvm/backtype/storm/scheduler/Cluster.java

```
@@ -0,0 +1,327 @@
+     * 
+     * @param cluster
+     * @return
+     */
+    public List<WorkerSlot> getAvailableSlots(SupervisorDetails supervisor) {
+        List<Integer> ports = this.getAvailablePorts(supervisor);
+        List<WorkerSlot> slots = new ArrayList<WorkerSlot>(ports.size());
+
+        for (Integer port : ports) {
+            slots.add(new WorkerSlot(supervisor.getId(), port));
+        }
+
+        return slots;
+    }
+    
+    /**
+     * get the unassigned executors of the topology.
+     */
+    public List<ExecutorDetails> getUnassignedExecutors(TopologyDetails topology) {
+        if (topology == null) {
+            return new ArrayList<ExecutorDetails>(0);
+        }
+
+        Collection<ExecutorDetails> allExecutors = topology.getExecutors();
+        SchedulerAssignment assignment = this.getAssignmentById(topology.getId());
+
+        if (assignment != null) {
+            Set<ExecutorDetails> assignedExecutors = assignment.getExecutors();
+            allExecutors.removeAll(assignedExecutors);
+        }
+
+        List<ExecutorDetails> ret = new ArrayList<ExecutorDetails>(allExecutors.size());
+        ret.addAll(allExecutors);
+
+        return ret;
+    }
+
+    /**
+     * Gets the number of workers assigned to this topology.
+     * 
+     * @param topology
+     * @return
+     */
+    public int getAssignedNumWorkers(TopologyDetails topology) {
+        SchedulerAssignment assignment = this.getAssignmentById(topology.getId());
+        if (topology == null || assignment == null) {
+            return 0;
+        }
+
+        Set<WorkerSlot> slots = new HashSet<WorkerSlot>();
+        slots.addAll(assignment.getExecutorToSlots().values());
+
+        return slots.size();
+    }
+
+    /**
+     * Assign the slot to the executors for this topology.
+     */
+    public void assign(WorkerSlot slot, String topologyId, Collection<ExecutorDetails> executors) {
+        SchedulerAssignment assignment = this.getAssignmentById(topologyId);
```

This function should throw an exception if the slot is already assigned.